### PR TITLE
Improve input validation and testing

### LINKED ISSUE #358: Clarify simple_image_name
`simple_image_name` in https://github.com/E3SM-Project/zppy/blob/main/tests/integration/utils.py needs to be fixed for proper testing.

Currently:
```
simple_image_name = image_name.split("/")[-1].split(".")[0]
```

For example both of
```
e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1850-1853/lat_lon/CERES-EBAF-surface-v4.1/ceres_ebaf_surface_v4.1-ALBEDO_SRF-SON-global.png
e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1850-1851/lat_lon/CERES-EBAF-surface-v4.1/ceres_ebaf_surface_v4.1-ALBEDO_SRF-DJF-global.png
```
would get mapped to `'ceres_ebaf_surface_v4'` in the `image_check_failures` directory, even though they have different years (`1850-1853` -- `1850-1851`) and different months (`SON` -- `DJF`)

### LINKED ISSUE #625: [Bug]: Failures on the Weekly test run
### What happened?

I originally ran the 3 weekly tests (bundles, comprehensive_v2, comprehensive_v3) after merging #604/#617 on 7/31. The intent was to re-run these tests every single week, but I thought it was reasonable to only run them on weeks where pull requests (with code changes rather than say doc changes) were merged into zppy. After all, if no changes had been merged, what would be the point of running the extremely lengthy tests? 

For #598, I was testing the latest sets (`tc_analysis`, `enso_diags`, `streamflow` [but apparently missing `qbo`]) added to the E3SM Diags CDAT migration (https://github.com/E3SM-Project/e3sm_diags/commits/cdat-migration-fy24), using  `min_case_e3sm_diags_cdat_migrated_sets`. However, these three sets didn't show up on the viewer.

Upon testing on `main`, using `weekly_comprehensive_v3`, I found that `tc_analysis` still wasn't plotting (though `enso_diags` and `streamflow` were). I then ran all the weekly tests yielding the following results:

### Sets rendered

| sub-task | sets included in viewer | sets that were specified but were not included |
| --- | --- | --- |
| e3sm_diags > [atm_monthly_180x360_aave](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-main-20240907/v3.LR.historical_0051/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1987-1988/viewer/) |  "lat_lon","enso_diags","diurnal_cycle","streamflow" | "tc_analysis","tropical_subseasonal" |
| e3sm_diags > [atm_monthly_180x360_aave_mvm](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-main-20240907/v3.LR.historical_0051/e3sm_diags/atm_monthly_180x360_aave_mvm/model_vs_model_1987-1988/viewer/) | "lat_lon", | N/A |
| e3sm_diags > [lnd_monthly_mvm_lnd](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-main-20240907/v3.LR.historical_0051/e3sm_diags/lnd_monthly_mvm_lnd/model_vs_model_1987-1988/viewer/) | "lat_lon_land" | N/A|

Why are `tc_analysis` and `tropical_subseasonal` missing in the rendering for model-vs-obs?
```
$ cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test-main-20240907/v3.LR.historical_0051/post/scripts/
$ grep -in tc_analysis e3sm_diags_atm_monthly_180x360_aave_model_vs_obs_1987-1988.o596145  
```
gives:
```
IndexError: list index out of range
```
and 
```
$ grep -in tropical_subseasonal e3sm_diags_atm_monthly_180x360_aave_model_vs_obs_1987-1988.o596145 
```
gives:
```
OSError: no files to open
```

Neither of these error messages are particularly enlightening.

### Image check failures
Test failed due to image check failures:
https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-main-20240907/v3.LR.historical_0051/image_check_failures_comprehensive_v3/

```
$ cd /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-main-20240907/v3.LR.historical_0051/image_check_failures_comprehensive_v3
$ ls *_diff* | wc -l
136
```
136 differences from the expected. Appears to be 7 MERRA2-related E3SM Diags diffs, 1 global-time-series diff, and 128 MPAS-Analysis diffs.

### Sets rendered

| sub-task | sets included in viewer | sets that were specified but were not included |
| --- | --- | --- |
| e3sm_diags > [atm_monthly_180x360_aave](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test-main-20240907/v2.LR.historical_0201/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1852-1853/viewer/) |  "lat_lon","diurnal_cycle","streamflow","tc_analysis" | "enso_diags" |
| e3sm_diags > [atm_monthly_180x360_aave_mvm](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test-main-20240907/v2.LR.historical_0201/e3sm_diags/atm_monthly_180x360_aave_mvm/model_vs_model_1852-1853/viewer/) | "lat_lon", | N/A |
| e3sm_diags > [lnd_monthly_mvm_lnd](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test-main-20240907/v2.LR.historical_0201/e3sm_diags/lnd_monthly_mvm_lnd/model_vs_model_1852-1853/viewer/) | "lat_lon_land" | N/A|

Why is `enso_diags` missing in the rendering for model-vs-obs?
```
$ cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v2_output/test-main-20240907/v2.LR.historical_0201/post/scripts/
$ grep -in enso_diags e3sm_diags_atm_monthly_180x360_aave_model_vs_obs_1852-1853.o596170
```
That does give:
```
IndexError: index 0 is out of bounds for axis 0 with size 0
RuntimeError: Requested years are outside of available sst obs records.
```
But didn't these years work before?? Why would they not now?

### Image check failures
Test failed due to image check failures:
https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test-main-20240907/v2.LR.historical_0201/image_check_failures_comprehensive_v2/

```
$ cd /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test-main-20240907/v2.LR.historical_0201/image_check_failures_comprehensive_v2
$ ls *_diff* | wc -l
12
```
12 differences from the expected, all of which are MERRA2-related E3SM Diags diffs

### Sets rendered

| sub-task | sets included in viewer | sets that were specified but were not included |
| --- | --- | --- |
| e3sm_diags > [atm_monthly_180x360_aave ](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_bundles_www/test-main-20240907/v3.LR.historical_0051/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1985-1986/viewer/)| "polar","enso_diags","diurnal_cycle", | N/A|
| e3sm_diags > [atm_monthly_180x360_aave_mvm](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_bundles_www/test-main-20240907/v3.LR.historical_0051/e3sm_diags/atm_monthly_180x360_aave_mvm/model_vs_model_1987-1988/viewer/) | "polar","enso_diags","streamflow",| "tc_analysis" |

Why is `tc_analysis` missing in the rendering for mvm?
```
$ cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_bundles_output/test-main-20240907/v3.LR.historical_0051/post/scripts
$ grep -in tc_analysis bundle3.o597964 
```
Not particularly enlightening error messages:
```
RuntimeError: Neither does AODMOM nor the variables in [('AODMOM',)] exist in the file
IndexError: list index out of range
```

### Image check failures
Test failed due to image check failures: 
https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_bundles_www/test-main-20240907/v3.LR.historical_0051/image_check_failures_bundles/

```
$ cd /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_bundles_www/test-main-20240907/v3.LR.historical_0051/image_check_failures_bundles
$ ls *_diff* | wc -l
1
```
1 difference from the expected: 1 global-time-series diff

## Possible reasons for image check failures

As far as I can recall, the expected results were generated after the 7/31 merging of #604/#617. There have been no code changes merged to `zppy` `main` since then. The only thing that could be different is the E3SM Diags environment used, but even that wouldn't account for the MPAS-Analysis diffs or global-time-series diffs. And even then, I was using `conda activate e3sm_diags_20240731`, so the diags environment should have been identical to when the expected results were generated.

Lesson learned: always run these tests on a weekly basis (even if no code changes have been merged!), to catch environmental changes e.g., build versions, new e3sm_diags/other package changes.

### What machine were you running on?

Chrysalis

### Environment

zppy `main` as of 9/24.

### What command did you run?

```bash
zppy -c tests/integration/generated/test_weekly_comprehensive_v3_chrysalis.cfg
zppy -c tests/integration/generated/test_weekly_comprehensive_v2_chrysalis.cfg
zppy -c tests/integration/generated/test_weekly_bundles_chrysalis.cfg # Run 1
zppy -c tests/integration/generated/test_weekly_bundles_chrysalis.cfg # Run 2 (for second part)
```

### Copy your cfg file

```bash
N/A
```

### What jobs are failing?

```bash
N/A
```

### What stack trace are you encountering?

```bash
N/A
```

### PR DESCRIPTION
**Pull request by forsyth2**
## Issue resolution
- Closes #625, #358

Select one: This pull request is...
- [ ] a bug fix: increment the patch version
- [x] a small improvement: increment the minor version
- [ ] an incompatible (non-backwards compatible) API change: increment the major version

## 1. Does this do what we want it to do?

Objectives:
Adding "defensive programming" techniques to catch bad input early.

Functional code:
- Added `guess_path_parameters` & `guess_section_parameters` parameters, defaulting to True, to keep backwards compatibility. Setting to False allows for more transparency of what paths are being used.
- Added required parameter checks for certain diags sets, based on default.ini notes. Also added `climo_diurnal_frequency` check.
- Added dependencies to .settings files
- Added `fail_on_dependency_skip` parameter, defaulting to False, to keep backwards compatibility. Setting to True stops zppy if any job launch fails.
- Refactored code to be more modular. This is a great improvement to both readability & testing. Previously, much of zppy had been written in monolithic functions that made testing outside of full integration testing difficult. Now, real unit testing (as in, independent of the outside world) is more feasible.
- Added exit to tc_analysis if the cyclone_stitch `dat` file is empty.

Testing:
- Run _all_ e3sm_diags sets on v2 and v3 comprehensive runs
- Changed reference years for e3sm_diags in v2 to start at 1870
- Added daily `ts` subtask to support the tropical_subseasonal e3sm_diags set.
- Remove `simple_image_name` so that _all_ image check failures actually show up in the web's diff directory. Resolves #358.
- Added unit tests for the newly modular code.
- Updated expected results for integration tests.

Required:
- [x] Product Management: I have confirmed with the stakeholders that the objectives above are correct and complete.
 - As zppy's lead developer, I am the stakeholder these objectives are most relevant for.
- [x] Testing: I have added or modified at least one "min-case" configuration file to test this change. Every objective above is represented in at least one cfg.
  - These are not necessary since the purpose of this PR is to fix the comprehensive Weekly integration testing & add more unit tests.
- [x] Testing: I have considered likely and/or severe edge cases and have included them in testing.
  - I have attempted exhaustive unit testing of functions that don't interact with the outside world (i.e., functions that wouldn't need to be wrapped in the `IO` monad were the code written in Haskell).

If applicable:
- [x] Testing: this pull request introduces an important feature or bug fix that we _must_ test often. I have updated the weekly-test configuration files, not just a "min-case" one.
- [x] Testing: this pull request adds at least one new possible parameter to the cfg. I have tested using this parameter with and without any other parameter that may interact with it.
  - `fail_on_dependency_skip` (`False` on v2,v3 tests; `True` on bundles tests) , `guess_path_parameters`, `guess_section_parameters` (`False` on v3 tests; `True` on v2, bundles tests)

## 2. Are the implementation details accurate & efficient?

Required:
- [x] Logic: I have visually inspected the entire pull request myself.
- [ ] Logic: I have left GitHub comments highlighting important pieces of code logic. I have had these code blocks reviewed by at least one other team member.

If applicable:
- [ ] Dependencies: This pull request introduces a new dependency. I have discussed this requirement with at least one other team member. The dependency is noted in `zppy/conda`, not just an `import` statement.
  - No new dependencies

## 3. Is this well documented?

Required:
- [x] Documentation: by looking at the docs, a new user could easily understand the functionality introduced by this pull request.
  - The bulk of `zppy` documentation is through reading parameter comments on `default.ini`.
  - This PR mainly affects developers, not users.

## 4. Is this code clean?

Required:
- [x] Readability: The code is as simple as possible and well-commented, such that a new team member could understand what's happening.
- [x] Pre-commit checks: All the pre-commits checks have passed.

If applicable:
- [ ] Software architecture: I have discussed relevant trade-offs in design decisions with at least one other team member. It is unlikely that this pull request will increase tech debt.

### CODE REVIEW

File: tests/integration/template_weekly_bundles.cfg [hunk 1]
```diff
@@ -95,11 +95,12 @@ years = "1985:1989:2",
   mapping_file = ""
   vars = "RIVER_DISCHARGE_OVER_LAND_LIQ"
 
-[tc_analysis]
-active = True
-bundle = "bundle3" # Let bundle1 finish first because "e3sm_diags: atm_monthly_180x360_aave_mvm" requires "ts: atm_monthly_180x360_aave"
-scratch = "#expand scratch#zppy_weekly_bundles_scratch/#expand unique_id#/#expand case_name#"
-years = "1985:1989:2",
+# TODO: Add "tc_analysis" back in after empty dat is resolved.
```
Comment (forsyth2) [inline tests/integration/template_weekly_bundles.cfg:98]: `tc_analysis` must be excluded for now because of https://github.com/E3SM-Project/e3sm_diags/pull/866#issuecomment-2405873119

File: tests/integration/utils.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 from mache import MachineInfo
 from PIL import Image, ImageChops, ImageDraw
 
-UNIQUE_ID = "unique_id"
+UNIQUE_ID = "test-pr628-20241011v12"
```
Comment (forsyth2) [inline tests/integration/utils.py:10]: TODO: switch back to `unique_id`

File: tests/integration/utils.py [hunk 3]
```diff
@@ -62,21 +62,20 @@ def compare_images(
 
             mismatched_images.append(image_name)
 
-            simple_image_name = image_name.split("/")[-1].split(".")[0]
             shutil.copy(
                 path_to_actual_png,
-                os.path.join(diff_dir, "{}_actual.png".format(simple_image_name)),
+                os.path.join(diff_dir, "{}_actual.png".format(image_name)),
```
Comment (forsyth2) [inline tests/integration/utils.py:67]: Resolves #358

File: zppy/templates/tc_analysis.bash [hunk 4]
```diff
@@ -90,6 +90,13 @@ cat ${result_dir}out.dat0* > ${result_dir}cyclones_${file_name}.txt
 StitchNodes --in_fmt "lon,lat,slp,wind" --in_connect ${result_dir}connect_CSne${res}_v2.dat --range 6.0 --mintime 6 --maxgap 1 --in ${result_dir}cyclones_${file_name}.txt --out ${result_dir}cyclones_stitch_${file_name}.dat --threshold "wind,>=,17.5,6;lat,<=,40.0,6;lat,>=,-40.0,6"
 rm ${result_dir}cyclones_${file_name}.txt
 
+# If cyclones_stitch file is empty, exit
+if ! [ -s ${result_dir}cyclones_stitch_${file_name}.dat ]; then
+  cd {{ scriptDir }}
+  echo 'ERROR (1)' > {{ prefix }}.status
+  exit 1
+fi
+
```
Comment (forsyth2) [inline zppy/templates/tc_analysis.bash:99]: Catches the error in https://github.com/E3SM-Project/e3sm_diags/pull/866#issuecomment-2405873119.

Comment (chengzhuzhang) [reply zppy/templates/tc_analysis.bash:99]: @forsyth2 I suggest not to catch this error, so that `e3sm_diags` runs won't be blocked by empty stitch files.

Comment (chengzhuzhang) [reply zppy/templates/tc_analysis.bash:99]: Based on your comment here: https://github.com/E3SM-Project/zppy/pull/632#issuecomment-2417430619.

Comment (forsyth2) [reply zppy/templates/tc_analysis.bash:99]: @chengzhuzhang I think it would be a mistake to let errors slide by silently, especially considering this one was hard to track down.

A couple options on the top of my mind:
1. Leave this as-is. If `tc_analysis` fails because of this, then users can remove `tc_analysis` from their list of sets in  the `e3sm_diags` task and then re-run `zppy`.
2. Have a parameter something like `fail_on_empty_stitch` to toggle this error on/off

Comment (chengzhuzhang) [reply zppy/templates/tc_analysis.bash:99]: My take is that this is not an error, just no cyclone is detected. Without catching the error, it should work more smoothly to print in the log that indicate no cycle stitch file is generated.

Comment (forsyth2) [reply zppy/templates/tc_analysis.bash:99]: That makes sense, but it _does_ cause an error in E3SM Diags down the road then. Notably, the `IndexError: list index out of range` error on https://github.com/E3SM-Project/zppy/issues/625#issuecomment-2401000188 didn't add much clarity to the error.

I think if we remove this error check in the `tc_analysis` bash script (which I agree we should do if it's not actually an error per-se), we should add an empty-dat check in the `e3sm_diags` task.
I've copied the stack trace again here:
```
Traceback (most recent call last):
  File "/home/ac.forsyth2/miniconda3/envs/e3sm_diags_1003/lib/python3.10/site-packages/e3sm_diags/parameter/core_parameter.py", line 266, in _run_diag
    single_result = module.run_diag(self)
  File "/home/ac.forsyth2/miniconda3/envs/e3sm_diags_1003/lib/python3.10/site-packages/e3sm_diags/driver/tc_analysis_driver.py", line 91, in run_diag
    test_data["metrics"] = generate_tc_metrics_from_te_stitch_file(test_te_file)
  File "/home/ac.forsyth2/miniconda3/envs/e3sm_diags_1003/lib/python3.10/site-packages/e3sm_diags/driver/tc_analysis_driver.py", line 172, in generate_tc_metrics_from_te_stitch_file
    te_stitch_vars = _get_vars_from_te_stitch(lines, max_len, num_storms)
  File "/home/ac.forsyth2/miniconda3/envs/e3sm_diags_1003/lib/python3.10/site-packages/e3sm_diags/driver/tc_analysis_driver.py", line 249, in _get_vars_from_te_stitch
    year_start = int(lines[0].split("\t")[2])
IndexError: list index out of range
```
```
    test_te_file = os.path.join(
        test_data_path,
        "cyclones_stitch_{}_{}_{}.dat".format(test_name, test_start_yr, test_end_yr),
    )
```
So, it would be nice to add a non-empty check to that `test_te_file`. Based on the stack trace, it would probably be easier to add that in `e3sm_diags` than in `zppy`.

Comment (chengzhuzhang) [reply zppy/templates/tc_analysis.bash:99]: Yes, it makes sense to catch it in e3sm_diags.

Comment (forsyth2) [reply zppy/templates/tc_analysis.bash:99]: Sounds good. And no, I don't think these `dat` files are used in any other sets. The TC analysis task isn't even a dependency for any other set:
```
    if "tc_analysis" in c["sets"]:
        dependencies.append(os.path.join(script_dir, f"tc_analysis{status_suffix}"))
```
